nengo_gui/sim.py

```diff
@@ -0,0 +1,436 @@
+
+    def load(self):
+        """Load the model and initialize everything"""
+        if self.filename == self.sim_server.filename:
+            # if we're on the default filenaem, just load it from the SimServer
+            self.model = self.sim_server.model
+            if self.sim_server.locals is None:
+                self.locals = None
+            else:
+                self.locals = self.sim_server.locals.copy()
+        else:
+            self.model = None
+            self.locals = None
+
+        # if we still don't have a locals dictionary, then run the script
+        if self.locals is None:
+            try:
+                with open(self.filename) as f:
+                    code = f.read()
+            except IOError:
+                code = ''
+
+            self.execute(code)
+
+        if self.model is None:
+            self.model = nengo.Network()
+            self.locals['model'] = self.model
+
+        # figure out good names for objects
+        self.name_finder = nengo_gui.NameFinder(self.locals, self.model)
+        self.default_labels = self.name_finder.known_name
+
+        # load the .cfg file
+        self.config = self.load_config()
+        self.config_save_needed = False
+        self.config_save_time = None   # time of last config file save
+
+        self.uid_prefix_counter = {}
+
+        self.create_components()
+
+    def create_components(self):
+        """Generate the actual Components from the Templates"""
+        self.components = []
+        self.template_uids = {}
+        for k, v in self.locals.items():
+            if isinstance(v, nengo_gui.components.component.Template):
+                self.template_uids[v] = k
+                c = v.create(self)
+                self.sim_server.component_uids[c.uid] = c
+                self.components.append(c)
+
+        # this ensures NetGraph, AceEditor, and SimControl are first
+        self.components.sort(key=lambda x: x.component_order)
+
+    def add_template(self, template):
+        """Add a new Component to an existing Sim."""
+        c = template.create(self)
+        self.sim_server.component_uids[c.uid] = c
+
+        self.components.append(c)
+
+        return c
+
+    def execute(self, code):
+        """Run the given code to generate self.model and self.locals.
+
+        The code will be stored in self.code, any output to stdout will
+        be a string as self.stdout, and any error will be in self.error.
+        """
+        locals = {}
+        locals['nengo_gui'] = nengo_gui
+        locals['__file__'] = self.filename
+
+        self.code = code
+        self.error = None
+        self.stdout = ''
+
+        patch = nengo_gui.monkey.Patch(self.filename)
+        try:
+            with patch:
+                exec(code, locals)
+        except nengo_gui.monkey.StartedSimulatorException:
+            # no running a simulator inside a script
+            pass
+        except nengo_gui.monkey.StartedVizException:
```

Comment:
Added warning messages!  We don't have a nice way of doing warnings (as opposed to errors), so I just put it the same place print statements go to.  I think in a future PR we could do something nicer with this, with a way to show debug and other logging statements....